src/lock_hash.py

```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env python3
+"""Charm code for MongoDB service."""
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
+import os
+from logging import getLogger
+
+import ops
+
+from config import Config
+from machine_helpers import ROOT_USER_GID
+
+logger = getLogger(__name__)
+
+HASH_KEY = "lockhash"
+LOCK_PATH = Config.MONGODB_DATA_DIR / f".{HASH_KEY}"
+UNDEFINED = "UNDEFINED"
+
+
+class LockHashHandler:
+    """Descriptor class for the lock hash stored in the file."""
+
+    def __set__(self, obj: ops.CharmBase, value: str):
```

Comment:
noob q: why obj and not unit? I would assume this could only be done on the unit?

Reply:
This is how data descriptors work in python.
They have to define at least two functions :  `__set__` and `__get__` which are called when accessing the descriptor or setting a value to it. They can define two other methods which are `__delete__` and  `__set_name__`.
The `__set__` function shall have three parameters which are `self`, the object owning the descriptor object in use and the value which is being written. Here it's the charm code that owns this objects (see `self.lock_hash = LockHashHandler()` in `src/charm.py` so it's the `obj` parameter.

`properties` are a very basic usage of descriptors (in the sense that they define a very basic descriptor.

If you want more information, I let you [read this descriptor guide](https://docs.python.org/3/howto/descriptor.html) from Python documentation :)

It seemed easier for me to export this in a descriptor class here in order to reduce the size of the charm class and make it easier to understand.